mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,295 @@
+from collections import OrderedDict
+from typing import Dict, List, Optional, Set, Tuple, cast
+
+from mypy.nodes import (
+    ARG_OPT, ARG_POS, MDEF, Argument, AssignmentStmt, Block, CallExpr, Context,
+    Decorator, Expression, JsonDict, NameExpr, SymbolTableNode, TempNode,
+    TypeInfo, Var
+)
+from mypy.plugin import ClassDefContext
+from mypy.plugins.common import _add_method, _get_decorator_bool_argument
+from mypy.types import CallableType, NoneTyp, Type, TypeVarDef, TypeVarType
+from mypy.typevars import fill_typevars
+
+# The set of decorators that generate dataclasses.
+dataclass_makers = {
+    'dataclass',
+    'dataclasses.dataclass',
+}
+
+
+class DataclassAttribute:
+    def __init__(
+            self,
+            name: str,
+            is_in_init: bool,
+            has_default: bool,
+            line: int, column: int,
+    ) -> None:
+        self.name = name
+        self.is_in_init = is_in_init
+        self.has_default = has_default
+        self.line = line
+        self.column = column
+
+    def to_argument(self, info: TypeInfo) -> Argument:
+        return Argument(
+            variable=self.to_var(info),
+            type_annotation=info[self.name].type,
+            initializer=None,
+            kind=ARG_OPT if self.has_default else ARG_POS,
+        )
+
+    def to_var(self, info: TypeInfo) -> Var:
+        return Var(self.name, info[self.name].type)
+
+    def serialize(self) -> JsonDict:
+        return {
+            'name': self.name,
+            'is_in_init': self.is_in_init,
+            'has_default': self.has_default,
+            'line': self.line,
+            'column': self.column,
+        }
```

Comment:
> The best I could come up with was to remove them from the class' symbol table and add the types to the DataclassAttributes but then I wasn't sure how to best serialize/deserialize the types.

You can add list of `InitVar`s here. Most types in mypy support (de-)serialization (those that are not supported like partial types, or forward reference types, should not appear in at the point where serialization happens). You can find signatures in `mypy.types.Type.serialize` and `mypy.types.Type.deserialize`.

Reply:
I had tried that, but deserialization failed with `De-serialization failure: TypeInfo not fixed`. Reading the docstring for `FakeInfo`, I understand this is supposed to be done in two passes, but it looks like the fixup step requires a lot of information I don't necessarily have/that it would be ugly to pass around. Am I missing something obvious?